enhancements/update/upgrade-without-registry.md

```diff
@@ -0,0 +1,887 @@
+          # Indicates if the images from the upgrade bundle have beel already
+          # copied to the `/etc/containers/storage` directory.
+          loaded: true|false
+          # Contains the metadata of the bundle as seen by this node. This is
+          # taken from the `metadata.json` file that is part of the bundle.
+          metadata: '{ ... }'
+        node1: { ... }
+        ...
+```
+
+And this is a summary of the new `ContainerRuntimeConfig` fields:
+
+```yaml
+apiVersion: machineconfiguration.openshift.io/v1
+kind: ContainerRuntimeConfig
+metadata:
+  name: pin-upgrade
+spec:
+  containerRuntimeConfig:
+    # Additional image store directories. Translates into something like this
+    # inside the `/etc/containers/storage.conf` file:
+    #
+    # additionalimagestores = [
+    #   "/var/lib/my-images/...",
+    #   "/var/lib/your-images/...",
+    #   ...
+    # ]
+    additionalImageStores:
+    - /var/lib/my-images
+    - /var/lib/my-images
+
+    # List of pinned images. Translates into a new `pin-upgrade.conf` file inside
+    # the `/etc/crio/crio.conf.d` directory with content similar to this:
+    #
+    # pinned_images = [
+    #   "quay.io/openshift-release-dev/ocp-release@sha256:...",
+    #   "quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:...",
+    #   "quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:...",
+    #   ...
+    # ]
+    #
+    # The CRI-O service will be reloaded when this is changed.
+    pinnedImages:
+    - quay.io/openshift-release-dev/ocp-release@sha256:...
+    - quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:...
+    - quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:...
+    ...
+```
+
+### Implementation Details/Notes/Constraints
+
+The proposed solution is based on copying all the required images to all the
+nodes of the cluster before starting the upgrade, and ensuring that no component
+requires access to a registry server during the upgrade. For this to work the
+following changes are required:
+
+1. No OpenShift component used during the upgrade should use the `Always` pull
+policy, as that forces the kubelet and CRI-O to try to contact the registry
+server even if the image is already available.
+
+1. No OpenShift component should garbage collect the upgrade images before or
+during the upgrade. This is typically started by the kubelet instruction CIR-O
+to remove images.
+
+1. No OpenShift component should explicitly try to contact the registry server
+without a fallback alternative.
+
+1. The engineer working in the partner's factory needs a `oc adm upgrade create
+bundle` tool to create the upgrade bundle.
+
+1. The technician needs documentation explaining how to use the upgrade bundle to
+perform the upgrade.
+
+1. The machine config operator should understand and support the changes that
+will be required to the CRI-O configuration during the upgrade.
+
+1. The cluster version operator needs to orchestrate the upgrade process.
+
```

Comment:
Also need to mention the need not to add/remove nodes while upgrading.

Reply:
In the "CVO will need to orchestrate the upgrade" we say this:

> To do so it will first disable auto-scaling to ensure that no new nodes
> are added to the cluster while this process is progress

Is that enough?